### PR TITLE
coding-conventions.chapter.md: update to account for #89885

### PR DESCRIPTION
**Pull request by ghost**
###### Description of changes

This PR: https://github.com/NixOS/nixpkgs/pull/89885 ensures that fetches are done securely (i.e. without `--insecure`) when the `hash` parameter is one of the four special "fake" hashes.  However the manual was not updated in that PR. This commit updates the manual to account for the already-merged changes from that PR.

ping @matthewbauer

###### Future-proofing

It would probably be a good idea if Nix didn't print out the complete unexpected hash of insecure fetches on the console, in order to deter people from copy-and-pasting it.  For example, by default the error message printed to the console could deliberately garble a few characters in the middle of the hash by replacing them with an invalid character like `.` unless `--option dammit-give-me-the-footgun` is set.

The question is how to communicate to Nix that the fetch was insecure.  I see two possibilities:

1. Nix adopts the nixpkgs convention that if expected-hash is `""`, `lib.fakeHash`, `lib.fakeSha256`, or `lib.fakeSha512` then the unexpected-hash was computed over something which was fetched securely.  This could be considered a layering violation.

2. A new derivation attribute `__insecure` for FODs is added, which tells Nix that the fetch was performed insecurely.  Since this would appear only on FODs it shouldn't affect any outpath addresses.  It should probably also be allowed on `__impure` derivations, and be inherited by FODs which depend on `__insecure __impure` derivations.

@edolstra if either of these approaches sounds acceptable to you I will implement it.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: doc/contributing/coding-conventions.chapter.md [hunk 1]
```diff
@@ -480,15 +480,10 @@ Preferred source hash type is sha256. There are several ways to get it.
 
 4. Extracting hash from local source tarball can be done with `sha256sum`. Use `nix-prefetch-url file:///path/to/tarball` if you want base32 hash.
 
-5. Fake hash: set fake hash in package expression, perform build and extract correct hash from error Nix prints.
-
-    For package updates it is enough to change one symbol to make hash fake. For new packages, you can use `lib.fakeSha256`, `lib.fakeSha512` or any other fake hash.
+5. Fake hash: set the hash to `""`, `lib.fakeHash`, `lib.fakeSha256`, or `lib.fakeSha512` in the package expression, perform build and extract correct hash from error Nix prints.  You must use one of these four fake hashes and not some randomly-chosen hash; see below for the reason.
```
Comment (fricklerhandwerk) [inline doc/contributing/coding-conventions.chapter.md:483]: Please do not refer to "below" or "above" in web documents. The document may change structure, and the verbal reference will be left dangling if change authors don't adapt it manually. People make mistakes, let's avoid opportunities for mistakes.

Lists of code tokens are easier to read as actual lists.

I think it was good to have a warning there, it just should be part of the list item.

```suggestion
5. Fake hash: set the hash to one of

   - `""`
   - `lib.fakeHash`
   - `lib.fakeSha256`
   - `lib.fakeSha512`
   
   in the package expression, attempt build and extract correct hash from error messages.

   :::{.warning}
   You must use one of these four fake hashes and not some arbitrarily-chosen hash.
   
   See [](#sec-source-hashes-security).
   :::
```

File: doc/contributing/coding-conventions.chapter.md [hunk 2]
```diff
@@ -500,7 +495,7 @@ Let's say Man-in-the-Middle (MITM) sits close to your network. Then instead of f
 
 - `https://` URLs are secure in methods 1, 2, 3;
 
-- `https://` URLs are not secure in method 5. When obtaining hashes with fake hash method, TLS checks are disabled. So refetch source hash from several different networks to exclude MITM scenario. Alternatively, use fake hash method to make Nix error, but instead of extracting hash from error, extract `https://` URL and prefetch it with method 1.
+- `https://` URLs are secure in method 5 *if* you use `""`, `lib.fakeHash`, `lib.fakeSha256`, or `lib.fakeSha512`.  If you use any other hash, `fetchurl` will pass `--insecure` to `curl` in order to be robust in the face of TLS certificate expiration and to allow fetching over protocols like HTTP which do not incur the overhead of TLS.
```
Comment (fricklerhandwerk) [inline doc/contributing/coding-conventions.chapter.md:498]: Let's emphasize this exception. Also refer to single source of truth, otherwise we will have to maintain that list twice.

Does the shorter wording still reflect what you wanted to say?

```suggestion
- `https://` URLs are secure in method 5 *only if* you use one of the listed fake hashes. If you use any other hash, `fetchurl` will pass `--insecure` to `curl` and may then degrade to HTTP in case of TLS certificate expiration.
```

**Review by fricklerhandwerk — CHANGES_REQUESTED** [review]

File: doc/contributing/coding-conventions.chapter.md (hunk 2 above)
Comment (amjoseph-nixpkgs) [reply doc/contributing/coding-conventions.chapter.md:498]: > Does the shorter wording still reflect what you wanted to say?

Yes, except for the assumption that this only affects HTTP (I referenced HTTP as an example).  I'll apply this and push another commit.

**Review by fricklerhandwerk — APPROVED** [review]
